accepted/2020/mono-convergence/globalization-support-in-.net5.md

```diff
@@ -0,0 +1,82 @@
+
+## .NET Globalization Invariant Mode
+There is already support for the globalization [invariant mode](https://github.com/dotnet/runtime/blob/master/docs/design/features/globalization-invariant-mode.md) since .NET Core 2.0. This will be the most size effective globalization mode for .NET5 which can be set by the developers. We are going to use existing logic and tweak it where necessary to allow IL Linker to remove any globalization code from the final app when running in invariant mode.
+
+## Size Sensitive Platforms
+Existing Xamarin platforms allow customers to choose which set of collation data they would like to add to their apps if any. This still makes sense for .NET5 as on some of the platform ICU dependency is not available from the system and we’ll need to bundle it to the final app. We could make this available to everyone but it’s most likely we will focus only on platforms which are size sensitive in .NET5 time frame.
+
+### Android
+Android has historically relied on ICU for some kind of globalization support.  From basically the beginning, `Locale` , `Character`, and some classes in `java.text` were provided in the platform globalization.  If you needed more than that, you could access `libicuuc` directly or choose to bundle another version in your APK.  That was pretty much the story up until API level 23.
+
+In API level 24 and higher, the platform white lists its version of `libicuuc` and also includes a subset of [ICU4J](https://developer.android.com/guide/topics/resources/internationalization#relation) API’s.  It’s important to note that [ICU4C](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/) (accessing `libicuuc` directly) is not part of the Android NDK.  What this means is that OEM’s are not required to follow any type of rules and they could limit / alter the data files as they see fit.  The potential for undefined behavior and additional testing on our part is very real. 
+
+Consult this [issue](https://github.com/android/ndk/issues/548#issuecomment-395561629) for detailed commentary on the challenges of making ICU an NDK API.
+
+### What Does This Mean For Us?
+The good news is that p-invoking from managed code works.  Assuming `libicuuc` continues to be white listed, this should not be a problem.  It would be wise if we explored using the `ICU4J`  API subset that’s exposed in the event they decide to tighten things up.  
+
+To summarize:
+
+* We can p-invoke into the system `libicuuc`.  Globalization for .net5 should work normally
+* Android 4.3 (API level 18) started bundling ICU 50 and our minimum version requirement is satisfied. 
+* OEM’s can potentially make life difficult by altering locales / globalization features
+* We should explore trying to interop with the `ICU4J` API and determine what the implications, if any are.
+
+### iOS/tvOS
```

Comment:
It would be useful to describe what Xamarin does today; and how our customers are happy with it.

Reply:
we ship with the tables that mono includes
we have some, optional `I18N.*.dll`, that customer can opt-in; and
we have some linker tricks to remove a few things (e.g. special calendars when the corresponding I18N support is not included)

Reply:
Customer wise there's not much negative feedback. Either they use Apple API or .net ones.

The most common issue is when mixing both. Apple design is different from .net one, e.g. `EN-fr` is valid on Apple platforms, but does not match to any .net `CultureInfo`. Not very hard to fix, but app developers must be aware of it.

Reply:
>It would be useful to describe what Xamarin does today

Xamarin globalization is not very compatible with .NET Core so this is not that relevant. We can decide iOS/tvOS (and not macOS) be the only platform where the globalization support is very different (and lacking full support for features like System.Globalization.Invariant mode) or unify. The existing implementation is a mix of native and managed pieces which use a subset of ICU data extracted some time ago. It also suffers from the suboptimal performance of string operations as the collation logic is fully managed with no advanced optimizations.

The most-reported bugs we get for globalization support are about inconsistent or missing data